crates/ruff_linter/src/rules/pylint/rules/unnecessary_chained_comparison.rs

```diff
@@ -0,0 +1,114 @@
+use std::collections::{HashMap, HashSet};
+
+use crate::checkers::ast::Checker;
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast};
+
+/// ## What it does
+/// Checks for boolean operations such as `a < b and b < c`
+/// that can be refactored into a single comparison `a < b < c`.
+///
+/// ## Why is this bad?
+/// A single comparison is semantically clearer and more concise.
+///
+/// ## Example
+/// ```python
+/// a = int(input())
+/// b = int(input())
+/// c = int(input())
+/// if a < b and b < c:
+///     pass
+/// ```
+///
+/// Use instead:
+/// ```python
+/// a = int(input())
+/// b = int(input())
+/// c = int(input())
+/// if a < b < c:
+///     pass
+/// ```
+
+#[violation]
+pub struct UnnecessaryChainedComparison;
+
+impl Violation for UnnecessaryChainedComparison {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Simplified chain comparison exists between the operands.")
+    }
+}
+
+// Each integer is a unique identifier for the node.
+#[derive(Default)]
+struct Bounds {
+    lower_bound: HashSet<i32>,
+    upper_bound: HashSet<i32>,
+}
```

Comment:
I'm having some trouble understanding the precise logic, but I wonder if it is worth paying the cost of allocating two hash sets for each `Bound`. Especially, what's unclear to me is if the uniqueness is important for the correctness of the rule. Ultimately, we only iterate over the nodes once. That means we should never end up inserting the same node twice into lower or upper bounds. 

I think we could instead use

* Use a `Vec<(u32, Bound)>` where `enum Bound { Lower, Upper }`
* We also know that the vec is sorted because we iterate the nodes in order. We can then implement a manual counting and use lookahead to to test if an id does both an upper and lower bound check

This has the downside that we might end up writing a few extra bounds if a comparison tests the same name multiple times but that's fine. The overhead for this is minimal.

Reply:
could avoid the allocation altogether for `bool_op.values.len() <= 32` by using bit sets for `lower_bound` and `upper_bound`

Reply:
Great idea around using a bit-set, let me know what you think: https://github.com/astral-sh/ruff/pull/11807/commits/eb08ee497eeeaa4d0ef78f80ba4192be0617a4ad